libs/xpdf/xpdf-src/doc/xpdf.1
.\" Copyright 1996-2025 Glyph & Cog, LLC
.TH xpdf 1 "14 Nov 2025"
.SH NAME
xpdf \- Portable Document Format (PDF) file viewer (version 4.06)
.SH SYNOPSIS
.B xpdf
[options]
.RI [ PDF-file
.RI [: page " | +" dest "]] ..."
.PP
.B xpdf
[options]
.B -remote
.IR remote-name " [" command " ...]"
.PP
.B xpdf
[options]
.B -open
.RI "[" PDF-file "]"
.SH DESCRIPTION
.B Xpdf
is a viewer for Portable Document Format (PDF) files.  (These are also
sometimes also called \'Acrobat' files, from the name of Adobe's PDF
software.)  Xpdf uses the Qt GUI toolkit and runs on Unix, OS X, and
Windows.
.PP
To run xpdf, type:
.PP
.RS
xpdf file.pdf
.RE
.PP
where
.I file.pdf
is your PDF file.  The file name can be followed by a page number to
be displayed, prefixed with a colon:
.PP
.RS
xpdf file.pdf :18
.RE
.PP
or by a named destination, prefixed with \'+' (this is only useful
with PDF files that provide named destination targets):
.PP
.RS
xpdf file.pdf +destinationA
.RE
.PP
If you specify multiple files, they will each be opened in a separate
tab:
.PP
.RS
xpdf file1.pdf file2.pdf :18 file3.pdf
.RE
.PP
You can also start xpdf without opening any files:
.PP
.RS
xpdf
.RE
.SH CONFIGURATION FILE
Xpdf reads a configuration file at startup.  It first tries to find
the user's private config file, ~/.xpdfrc.  If that doesn't exist, it
looks for a system-wide config file, typically /etc/xpdfrc (but this
location can be changed when xpdf is built).  See the
.BR xpdfrc (5)
man page for details.
.SH OPTIONS
The following command line options are available.  All command line
options must come before any PDF files to be opened.
.PP
Many of the options can be set with configuration file commands.
These are listed in square brackets with the description of the
corresponding command line option.
.TP
.BI \-geometry " geometry"
Set the initial window geometry.
.TP
.BI \-title " title"
Set the window title.  By default, the title will be "xpdf: foo.pdf".
.TP
.BI \-open " \fR[\fPPDF-file\fR]\fP"
This option sets up a default remote server.  If Xpdf is already
running (with the "-open" switch), the PDF file (if any) is opened in
a new tab.  If Xpdf (with the "-open" switch) is not already running,
starts Xpdf and opens the PDF file (if any).  This is useful for GUI
desktop environments, e.g., the typical double-click on a PDF file
case.
.TP
.B \-rv
Set reverse video mode.  This reverses the colors of everything except
images.  It may not always produce great results for PDF files which
do weird things with color.  This also causes the paper color to
default to black.
.TP
.BI \-papercolor " color"
Set the "paper color", i.e., the background of the page display.  The
color can be #RRGGBB (hexadecimal) or a named color.  This option will
not work well with PDF files that do things like filling in white
behind the text.
.RB "[config file: " paperColor ]
.TP
.BI \-mattecolor " color"
Set the matte color, i.e., the color used for background outside the
actual page area.  The color can be #RRGGBB (hexadecimal) or a named
color.
.RB "[config file: " matteColor ]
.TP
.BI \-fsmattecolor " color"
Set the matte color for full-screen mode.  The color can be #RRGGBB
(hexadecimal) or a named color.
.RB "[config file: " fullScreenMatteColor ]
.TP
.BI \-z " zoom"
Set the initial zoom factor.  A number specifies a zoom percentage,
where 100 means 72 dpi.  You may also specify \'page', to fit the page
to the window size, or \'width', to fit the page width to the window
width.
.RB "[config file: " initialZoom ]
.TP
.BI \-rot " angle"
Rotate pages by 0 (the default), 90, 180, or 270 degrees.
.TP
.BI \-aa " yes | no"
Enable or disable font anti-aliasing.  This defaults to "yes".
.RB "[config file: " antialias ]
.TP
.BI \-aaVector " yes | no"
Enable or disable vector anti-aliasing.  This defaults to "yes".
.RB "[config file: " vectorAntialias ]
.TP
.BI \-enc " encoding-name"
Sets the encoding to use for text output.  The
.I encoding\-name
must be defined with the unicodeMap command (see
.BR xpdfrc (5)).
This defaults to "Latin1" (which is a built-in encoding).
.RB "[config file: " textEncoding ]
.TP
.BI \-pw " password"
Specify the password for the PDF file.  This can be either the owner
password (which will bypass all security restrictions) or the user
password.
.TP
.B \-fullscreen
Open xpdf in full-screen mode, useful for presentations.
.TP
.BI \-remote " remote-name"
Start Xpdf in remote server mode.  See the REMOTE SERVER MODE section.
.TP
.BI \-display " display"
Set the X display (only available with X11).
.TP
.B \-cmd
Print commands as they're executed (useful for debugging).
.RB "[config file: " printCommands ]
.TP
.BI \-tabstate " tabstate-file"
Sets the file used by the loadTabState and saveTabState commands.
.RB "[config file: " tabStateFile ]
.TP
.BI \-cfg " config-file"
Read
.I config-file
in place of ~/.xpdfrc or the system-wide config file.
.TP
.B \-v
Print copyright and version information.
.TP
.B \-h
Print usage information.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.PP
.SH CONTROLS
.SS Toolbar
.TP
.B "toggle sidebar button"
Toggles (i.e., shows or hides) the sidebar.
.TP
.B "status indicator"
This icon is animated while Xpdf is rendering a page.  It turns red
when an error or warning has been issued.  Clicking on it opens the
error dialog.
.TP
.B "selection mode"
This icon is an "I-beam" in linear selection mode, and an arrow in
block selection mode.  Clicking on it toggles between the two
selection modes.
.TP
.B "page number entry box"
Move to a specific page number.  Click in the box to activate it, type
the page number, then hit return.  This will instead display and
accept page labels, if the "view - page labels" menu item is checked.
.TP
.B "left/right arrow buttons"
Go backward or forward along the history path.
.TP
.B "zoom out/in buttons"
Zoom out or in (i.e., change magnification) incrementally.
.TP
.B "zoom popup menu"
Change the zoom factor (see the description of the \-z option above).
.TP
.B "fit width button"
Change the zoom factor to fit the page width to the window width.
.TP
.B "fit page button"
Change the zoom factor to fit the page to the window size.
.TP
.B "find entry box"
Find a text string.  Click in the box to activate it, type a search
string, then hit return.
.TP
.B "find next button"
Find the next occurrence of the search string.
.TP
.B "find previous button"
Find the previous occurrence of the search string.
.TP
.B "find settings button"
Display the current find settings: case insensitive, case sensitive,
smart case (case sensitive only if the search string contains
uppercase characters), find whole words.
.PP
.SS Menu bar
The menu bar is above the toolbar.  The menu items should be
self-explanatory.
.PP
.SS Tab list
The tab list is on the left, just below the toolbar.  It lists all
open tabs.
.PP
.SS Outline/layers/attachments pane
This pane is on the left, below the tab list.  The popup allows you to
select from outline, layers, or attachments.
.PP
The outline is a tree-like structure of bookmarks that allows moving
within the PDF file.  Not all PDF files have outlines.
.PP
Layers (a.k.a. optional content) allow parts of the PDF content to be
shown or hidden.  Not all PDF files have layers.
.PP
Attachments are other files embedded within the PDF file.  There is
a 'save' button for each attached file.  Not all PDF files have
attachments.
.PP
.SS Text selection
In block selection mode, dragging the mouse with the left button held
down will highlight an arbitrary rectangle.  Shift-clicking will
extend the selection.
.PP
In linear selection mode, dragging with the left button will highlight
text in reading order.  Double-clicking or triple-clicking will select
a word or a line, respectively.  Shift-clicking will extend the
selection.
.PP
Selected text can be copied to the clipboard (with the edit/copy menu
item).  On X11, selected text will be available in the X selection
buffer.
.PP
.SS Links
When the mouse is over a hyperlink, the link target will be shown in a
popup near the bottom of the window.
.PP
Clicking on a hyperlink will jump to the link's destination.  A link
to another PDF document will make xpdf load that document.  A
\'launch' link to an executable program will display a dialog, and if
you click \'ok', execute the program.  URL links are opened in a
system-dependent way.  (On UNIX, Qt uses the $BROWSER environment
variable.)
.PP
.SS Mouse bindings
The left mouse button is used to select text (see above).
.PP
Clicking on a link with the middle button opens the link in a new tab.
.PP
Dragging the mouse with the middle button held down pans the window.
.PP
The right mouse button opens a popup menu (see
.BR popupMenuCmd " in " xpdfrc (5)).
.PP
.SS Key bindings
This section lists the default key bindings.  Bindings can be changed
using the config file (see
.BR xpdfrc (5)).
.TP
.B control-o
Open a new PDF file via a file requester.
.TP
.B control-r
Reload the current PDF file.  Note that Xpdf will reload the file
automatically (on a page change or redraw) if it has changed since it
was last loaded.
.TP
.B control-f
Find a text string.  This sets keyboard focus to the \'find' box.
.TP
.B control-G
Find next occurrence.
.TP
.B control-C
Copy selected text to the clipboard.
.TP
.B control-P
Print.
.TP
.BR control-0 " (control-zero)"
Set the zoom factor to 125%.
.TP
.BR control-+ " (control-plus)"
Zoom in (increment the zoom factor by 1).
.TP
.BR control-- " (control-minus)"
Zoom out (decrement the zoom factor by 1).
.TP
.B control-s
Save PDF via a file requester.
.TP
.B control-t
Open a new tab.
.TP
.B control-n
Open a new window.
.TP
.B control-w
Close the current tab.  Closes the window if this was the last open
tab.  Quits the application if this was the last open window.
.TP
.B control-l
Toggle between full-screen and window modes.
.TP
.B control-q
Quit.
.TP
.B control-<tab>
Next tab.
.TP
.B control-shift-<tab>
Previous tab.
.TP
.B control-?
Help.
.TP
.B alt-<left-arrow>
Go backward along the history path.
.TP
.B alt-<right-arrow>
Go forward along the history path.
.TP
.B home
Go to the top left of current page.
.TP
.B control-<home>
Go to the first page.
.TP
.B end
Go to the bottom right of current page.
.TP
.B control-<end>
Go to the last page.
.TP
.BR <space> " or " <PageDown>
Scroll down on the current page; if already at bottom, move to next
page.
.TP
.BR control-<PageDown> " or " control-<down-arrow>
Go to the next page.  If <ScrollLock> is active, this maintains the
relative position on the page.
.TP
.B <PageUp>
Scroll up on the current page; if already at top, move to previous
page.
.TP
.BR control-<PageUp> " or " control-<up-arrow>
Go to the previous page.  If <ScrollLock> is active, this maintains
the relative position on the page.
.TP
.B <esc>
Exit full-screen mode.
.TP
.B arrows
Scroll the current page.
.TP
.B g
Set keyboard focus to the page number entry box.
.TP
.B z
Set the zoom factor to \'page' (fit page to window).
.TP
.B w
Set the zoom factor to \'width' (fit page width to window).
.SH Full-screen mode
Xpdf can be placed into full-screen mode via the
.B \-fullscreen
command line option, the \'full screen' menu item, or a binding to the
.B fullScreenMode
or
.B toggleFullScreenMode
command.
.PP
Entering full-screen mode automatically switches to single-page view
mode and to the fit-page zoom factor.
.PP
Full-screen mode can be exited via the default <esc> key binding,
or via a binding to the
.B windowMode
or
.BR toggleFullScreenMode command .
.SH COMMANDS
Xpdf's key and mouse bindings are user-configurable, using the bind
and unbind commands in the config file (see
.BR xpdfrc (5)).
The bind command allows you to bind a key or mouse button to a
sequence of one or more commands.
.PP
In commands that take arguments (inside parentheses), special
characters (namely '(', ')', ',', and '\\x01') can be escaped by
preceding them with a '\\x01' character.  This is mostly useful in
things like scripts that need to be able to open arbitrary PDF files,
using the \'openFile' command.
.PP
The following commands are supported:
.TP
.B about
Open the \'about' dialog.
.TP
.B blockSelectMode
Set block selection mode.  In this mode, the selection is a simple
rectangle.  Any part of the page can be selected, regardless of the
content on the page.
.TP
.BI checkOpenFile( file )
Check that
.I file
is open in the current tab, and open it if not.
.TP
.BI checkOpenFileAtDest( file, dest )
Check that
.I file
is open in the current tab, and open it if not.  In either case go to
the specified named destination.
.TP
.BI checkOpenFileAtPage( file, page )
Check that
.I file
is open in the current tab, and open it if not.  In either case go to
the specified page.
.TP
.B closeSidebar
Close the sidebar.
.TP
.B closeSidebarMoveResizeWin
Close the sidebar, resizing the window so that the document size
doesn't change, and moving the window so that the document stays in
the same place on the screen.
.TP
.B closeSidebarResizeWin
Close the sidebar, resizing the window so that the document size
doesn't change.
.TP
.B closeTabOrQuit
Close the tab.  If this was the last open tab, close the window.  If
this was the last window open, quit.
.TP
.B closeWindowOrQuit
Close the window.  If this was the last open window, quit.
.TP
.B continuousMode
Switch to continuous view mode.
.TP
.B copy
Copy selected text to the clipboard.
.TP
.B copyLinkTarget
Copy the target of the link under the mouse cursor to the clipboard.
.TP
.B endPan
End a pan operation.
.TP
.B endSelection
End a selection.
.TP
.BI expandSidebar( n )
Expand the sidebar by
.I n
pixels.  Opens the sidebar if it is currently closed.
.TP
.B find
Set keyboard focus to the \'find' box.
.TP
.B findFirst
Find the first occurrence of the search string.
.TP
.B findNext
Find the next occurrence of the search string.
.TP
.B findPrevious
Find the previous occurrence of the search string.
.TP
.B focusToDocWin
Set keyboard focus to the main document window.
.TP
.B focusToPageNum
Set keyboard focus to the page number text box.
.TP
.B followLink
Follow a hyperlink (does nothing if the mouse is not over a link).
.TP
.B followLinkInNewTab
Follow a hyperlink, opening PDF files in a new tab (does nothing if
the mouse is not over a link).  For links to non-PDF files, this
command is identical to followLink.
.TP
.B followLinkInNewTabNoSel
Same as followLinkInNewTab, but does nothing if there is a non-empty
selection.  (This is useful as a mouse button binding.)
.TP
.B followLinkInNewWin
Follow a hyperlink, opening PDF files in a new window (does nothing if
the mouse is not over a link).  For links to non-PDF files, this
command is identical to followLink.
.TP
.B followLinkInNewWinNoSel
Same as followLinkInNewWin, but does nothing if there is a non-empty
selection.  (This is useful as a mouse button binding.)
.TP
.B followLinkNoSel
Same as followLink, but does nothing if there is a non-empty selection.
(This is useful as a mouse button binding.)
.TP
.B fullScreenMode
Go to full-screen mode.
.TP
.B goBackward
Move backward along the history path.
.TP
.B goForward
Move forward along the history path.
.TP
.BI gotoDest( dest )
Go to a named destination.
.TP
.B gotoLastPage
Go to the last page in the PDF file.
.TP
.BI gotoPage( page )
Go to the specified page.
.TP
.BI help
Open the help URL.
.TP
.B hideMenuBar
Hide the menu bar.
.TP
.B hideToolbar
Hide the toolbar.
.TP
.B horizontalContinuousMode
Switch to horizontal continuous view mode.
.TP
.B linearSelectMode
Set linear selection mode.  In this mode, the selection follows text.
Non-text regions cannot be selected.
.TP
.B loadSession
Load the session from the session save file.  The path for the session file is specified with the sessionFile setting (see
.BR xpdfrc (5)).
.TP
.BI loadTabState
Load the tab state file (which was written via the saveTabState
command), and restore the tabs listed in that file.  The path for the
tab state file is specified with the tabStateFile setting (see
.BR xpdfrc (5)).
.TP
.B newTab
Open an empty new tab.
.TP
.B newWindow
Open an empty new window.
.TP
.B nextPage
Go to the next page.
.TP
.B nextPageNoScroll
Go to the next page, with the current relative scroll position.
.TP
.B nextTab
Switch to the next tab.
.TP
.B normalVideoMode
Switch to normal video mode (the opposite of reverse video mode).
.TP
.B open
Open a PDF file in the current tab, using the open dialog.
.TP
.B openErrorWindow
Open the error window.
.TP
.BI openFile( file )
Open the specified file in the current tab.
.TP
.BI openFile2( file, page, dest, passwd, location )
Open the specified file.  If
.I dest
is not empty, go to the specified named destination.  Else, if
.I page
is not empty, go to the specified page number.  If
.I password
is not empty, it is used as the PDF password.  If
.I location
is "win", open the file in a new window; if it is "tab", open in a new
tab; if it is "check", open in the current tab, but only if the
specified PDF file isn't already open; else open in the current tab.
Any/all of the arguments, other than
.IR file ,
can be empty strings.  For example:
.nf
openFile2(test.pdf,7,,,tab)
.fi
.TP
.BI openFileAtDest( file, dest )

ChangeLog
2025-10-16  Karl Berry  <karl@freefriends.org>

	* Build (status): avoid using egrep.
	Report from Jan Kasprzak,
	https://tug.org/pipermail/tex-k/2025-October/004217.html.

2025-09-07  Karl Berry  <karl@freefriends.org>

	* Makefile.am (install_target): new variable, install-strip by default.
	(world): use it.

2025-07-22  Karl Berry  <karl@freefriends.org>

	* reautoconf: mention that autoconf-only packages need to be
	explicitly added; would be better to read kpse-pkgs.m4.
	Also reautoconf tree for kpse-setup.m4 changes.

2025-06-26  Karl Berry  <karl@freefriends.org>

	* automake 1.18.1.

2025-04-30  Karl Berry  <karl@freefriends.org>

	* Build: incorporate original CFLAGS, CXXFLAGS, OBJCXXFLAGS values.
	https://tug.org/pipermail/tlbuild/2025q2/005771.html

2025-04-29  Karl Berry  <karl@freefriends.org>

	* Build: use TL_CFLAGS, TL_CXXFLAGS, etc., for TL_BUILD_ENV
	if -g is not given.

2025-03-08  Karl Berry  <karl@tug.org>

	* version.ac: 2026/dev.

2025-03-07  Karl Berry  <karl@tug.org>

	* TL'25 release.

2025-01-12  Karl Berry  <karl@tug.org>

	* Build: also support --debug-more for -g -Og -ggdb3,
	per Luigi, as described on https://tug.org/texlive/build.html.

2024-07-12  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake 1.17.

2024-04-29  Karl Berry  <karl@freefriends.org>

	* Makefile.am (distdir): override default dist generation,
	since we don't try to keep make dist working any more.
	https://tug.org/pipermail/tlbuild/2024q2/005547.html

2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the

libs/xpdf/xpdf-3.04/xpdf/xpdf.cc
//========================================================================
//
// xpdf.cc
//
// Copyright 1996-2003 Glyph & Cog, LLC
//
//========================================================================

#include <aconf.h>
#include "gtypes.h"
#include "GString.h"
#include "parseargs.h"
#include "gfile.h"
#include "gmem.h"
#include "GlobalParams.h"
#include "Object.h"
#include "XPDFApp.h"
#include "config.h"

//------------------------------------------------------------------------
// command line options
//------------------------------------------------------------------------

static GBool contView = gFalse;
static char enableFreeTypeStr[16] = "";
static char antialiasStr[16] = "";
static char vectorAntialiasStr[16] = "";
static char psFileArg[256];
static char paperSize[15] = "";
static int paperWidth = 0;
static int paperHeight = 0;
static GBool level1 = gFalse;
static char textEncName[128] = "";
static char textEOL[16] = "";
static char ownerPassword[33] = "\001";
static char userPassword[33] = "\001";
static GBool fullScreen = gFalse;
static char remoteName[100] = "xpdf_";
static char remoteCmd[512] = "";
static GBool doRemoteReload = gFalse;
static GBool doRemoteRaise = gFalse;
static GBool doRemoteQuit = gFalse;
static GBool printCommands = gFalse;
static GBool quiet = gFalse;
static char cfgFileName[256] = "";
static GBool printVersion = gFalse;
static GBool printHelp = gFalse;

static ArgDesc argDesc[] = {
  {"-g",          argStringDummy, NULL,           0,
   "initial window geometry"},
  {"-geometry",   argStringDummy, NULL,           0,
   "initial window geometry"},
  {"-title",      argStringDummy, NULL,           0,
   "window title"},
  {"-cmap",       argFlagDummy,   NULL,           0,
   "install a private colormap"},
  {"-rgb",        argIntDummy,    NULL,           0,
   "biggest RGB cube to allocate (default is 5)"},
  {"-rv",         argFlagDummy,   NULL,           0,
   "reverse video"},
  {"-papercolor", argStringDummy, NULL,           0,
   "color of paper background"},
  {"-z",          argStringDummy, NULL,           0,
   "initial zoom level (percent, 'page', 'width')"},
  {"-cont",       argFlag,        &contView,      0,
   "start in continuous view mode" },
#if HAVE_FREETYPE_FREETYPE_H | HAVE_FREETYPE_H
  {"-freetype",   argString,      enableFreeTypeStr, sizeof(enableFreeTypeStr),
   "enable FreeType font rasterizer: yes, no"},
#endif
  {"-aa",         argString,      antialiasStr,   sizeof(antialiasStr),
   "enable font anti-aliasing: yes, no"},
  {"-aaVector",   argString,      vectorAntialiasStr, sizeof(vectorAntialiasStr),
   "enable vector anti-aliasing: yes, no"},
  {"-ps",         argString,      psFileArg,      sizeof(psFileArg),
   "default PostScript file name or command"},
  {"-paper",      argString,      paperSize,      sizeof(paperSize),
   "paper size (letter, legal, A4, A3, match)"},
  {"-paperw",     argInt,         &paperWidth,    0,
   "paper width, in points"},
  {"-paperh",     argInt,         &paperHeight,   0,
   "paper height, in points"},
  {"-level1",     argFlag,        &level1,        0,
   "generate Level 1 PostScript"},
  {"-enc",    argString,   textEncName,    sizeof(textEncName),
   "output text encoding name"},
  {"-eol",    argString,   textEOL,        sizeof(textEOL),
   "output end-of-line convention (unix, dos, or mac)"},
  {"-opw",        argString,      ownerPassword,  sizeof(ownerPassword),
   "owner password (for encrypted files)"},
  {"-upw",        argString,      userPassword,   sizeof(userPassword),
   "user password (for encrypted files)"},
  {"-fullscreen", argFlag,        &fullScreen,    0,
   "run in full-screen (presentation) mode"},
  {"-remote",     argString,      remoteName + 5, sizeof(remoteName) - 5,
   "start/contact xpdf remote server with specified name"},
  {"-exec",       argString,      remoteCmd,      sizeof(remoteCmd),
   "execute command on xpdf remote server (with -remote only)"},
  {"-reload",     argFlag,        &doRemoteReload, 0,
   "reload xpdf remote server window (with -remote only)"},
  {"-raise",      argFlag,        &doRemoteRaise, 0,
   "raise xpdf remote server window (with -remote only)"},
  {"-quit",       argFlag,        &doRemoteQuit,  0,
   "kill xpdf remote server (with -remote only)"},
  {"-cmd",        argFlag,        &printCommands, 0,
   "print commands as they're executed"},
  {"-q",          argFlag,        &quiet,         0,
   "don't print any messages or errors"},
  {"-cfg",        argString,      cfgFileName,    sizeof(cfgFileName),
   "configuration file to use in place of .xpdfrc"},
  {"-v",          argFlag,        &printVersion,  0,
   "print copyright and version info"},
  {"-h",          argFlag,        &printHelp,     0,
   "print usage information"},
  {"-help",       argFlag,        &printHelp,     0,
   "print usage information"},
  {"--help",  argFlag,     &printHelp,     0,
   "print usage information"},
  {"-?",      argFlag,     &printHelp,     0,
   "print usage information"},
  {NULL}
};

//------------------------------------------------------------------------

int main(int argc, char *argv[]) {
  XPDFApp *app;
  GString *fileName;
  int pg;
  GString *destName;
  GString *userPasswordStr, *ownerPasswordStr;
  GBool ok;
  int exitCode;

  exitCode = 0;
  userPasswordStr = ownerPasswordStr = NULL;

  // parse args
  ok = parseArgs(argDesc, &argc, argv);
  if (!ok || printVersion || printHelp) {
    fprintf(stderr, "xpdf version %s\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    if (!printVersion) {
      printUsage("xpdf", "[<PDF-file> [<page> | +<dest>]]", argDesc);
    }
    exitCode = 99;
    goto done0;
  }

  // read config file
  globalParams = new GlobalParams(cfgFileName);
  globalParams->setupBaseFonts(NULL);
  if (contView) {
    globalParams->setContinuousView(contView);
  }
  if (psFileArg[0]) {
    globalParams->setPSFile(psFileArg);
  }
  if (paperSize[0]) {
    if (!globalParams->setPSPaperSize(paperSize)) {
      fprintf(stderr, "Invalid paper size\n");
    }
  } else {
    if (paperWidth) {
      globalParams->setPSPaperWidth(paperWidth);
    }
    if (paperHeight) {
      globalParams->setPSPaperHeight(paperHeight);
    }
  }
  if (level1) {
    globalParams->setPSLevel(psLevel1);
  }
  if (textEncName[0]) {
    globalParams->setTextEncoding(textEncName);
  }
  if (textEOL[0]) {
    if (!globalParams->setTextEOL(textEOL)) {
      fprintf(stderr, "Bad '-eol' value on command line\n");
    }
  }
  if (enableFreeTypeStr[0]) {
    if (!globalParams->setEnableFreeType(enableFreeTypeStr)) {
      fprintf(stderr, "Bad '-freetype' value on command line\n");
    }
  }
  if (antialiasStr[0]) {
    if (!globalParams->setAntialias(antialiasStr)) {
      fprintf(stderr, "Bad '-aa' value on command line\n");
    }
  }
  if (vectorAntialiasStr[0]) {
    if (!globalParams->setVectorAntialias(vectorAntialiasStr)) {
      fprintf(stderr, "Bad '-aaVector' value on command line\n");
    }
  }
  if (printCommands) {
    globalParams->setPrintCommands(printCommands);
  }
  if (quiet) {
    globalParams->setErrQuiet(quiet);
  }

  // create the XPDFApp object
  app = new XPDFApp(&argc, argv);

  // the initialZoom parameter can be set in either the config file or
  // as an X resource (or command line arg)
  if (app->getInitialZoom()) {
    globalParams->setInitialZoom(app->getInitialZoom()->getCString());
  }

  // check command line
  ok = ok && argc >= 1 && argc <= 3;
  if (remoteCmd[0]) {
    ok = ok && remoteName[5] && !doRemoteReload && !doRemoteRaise &&
         !doRemoteQuit && argc == 1;
  }
  if (doRemoteReload) {
    ok = ok && remoteName[5] && !doRemoteQuit && argc == 1;
  }
  if (doRemoteRaise) {
    ok = ok && remoteName[5] && !doRemoteQuit;
  }
  if (doRemoteQuit) {
    ok = ok && remoteName[5] && argc == 1;
  }
  if (!ok || printVersion || printHelp) {
    fprintf(stderr, "xpdf version %s\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    if (!printVersion) {
      printUsage("xpdf", "[<PDF-file> [<page> | +<dest>]]", argDesc);
    }
    exitCode = 99;
    goto done1;
  }
  if (argc >= 2) {
    fileName = new GString(argv[1]);
  } else {
    fileName = NULL;
  }
  pg = 1;
  destName = NULL;
  if (argc == 3) {
    if (argv[2][0] == '+') {
      destName = new GString(&argv[2][1]);
    } else {
      pg = atoi(argv[2]);
      if (pg < 0) {
	fprintf(stderr, "Invalid page number (%d)\n", pg);
	exitCode = 99;
	goto done2;
      }
    }
  }

  // handle remote server stuff
  if (remoteName[5]) {
    app->setRemoteName(remoteName);
    if (app->remoteServerRunning()) {
      if (fileName) {
	if (destName) {
	  app->remoteOpenAtDest(fileName, destName, doRemoteRaise);
	} else {
	  app->remoteOpen(fileName, pg, doRemoteRaise);
	}
      } else if (remoteCmd[0]) {
	app->remoteExec(remoteCmd);
      } else if (doRemoteReload) {
	app->remoteReload(doRemoteRaise);
      } else if (doRemoteRaise) {
	app->remoteRaise();
      } else if (doRemoteQuit) {
	app->remoteQuit();
      }
      goto done2;
    }
    if (doRemoteQuit) {
      goto done2;
    }
  }

  // set options
  app->setFullScreen(fullScreen);

  // check for password string(s)
  ownerPasswordStr = ownerPassword[0] != '\001' ? new GString(ownerPassword)
                                                : (GString *)NULL;
  userPasswordStr = userPassword[0] != '\001' ? new GString(userPassword)
                                              : (GString *)NULL;

  // open the file and run the main loop
  if (destName) {
    if (!app->openAtDest(fileName, destName,
			 ownerPasswordStr, userPasswordStr)) {
      exitCode = 1;
      goto done2;
    }
  } else {
    if (!app->open(fileName, pg, ownerPasswordStr, userPasswordStr)) {
      exitCode = 1;
      goto done2;
    }
  }
  app->run();

  exitCode = 0;

  // clean up
 done2:
  if (userPasswordStr) {
    delete userPasswordStr;
  }
  if (ownerPasswordStr) {
    delete ownerPasswordStr;
  }
  if (destName) {
    delete destName;
  }
  if (fileName) {
    delete fileName;
  }
 done1:
  delete app;
  delete globalParams;

  // check for memory leaks
 done0:
  Object::memCheck(stderr);
  gMemReport(stderr);

  return exitCode;
}
